### PR TITLE
[Final Project] Facade_Layout_Compiler

### LINKED ISSUE #618: Project Proposal: Facade Layout Compiler
**What will you do?**

Image ← This is an abstract facade layout diagram.
row 1: E E E E E E E E
row 2: E E E E E E E E
row 3: E W S S S S W E
row 4: E W S S S S W E
row 5: E E E D D E E E

I will design and implement a Facade Layout Compiler — a domain-specific compiler that parses a simple textual description of a building façade (a grid of doors, windows, and spacing) and compiles it into maybe both visual (SVG) and structured (JSON) outputs.
The compiler will check architectural layout rules such as door placement, window spacing, column alignment, and façade symmetry.
In short, it turns high-level façade specifications into verified, renderable layouts through standard compiler stages.

**How will you do it?**
The project will be implemented in Python. The compiler will have:
Front-end: A lexer and parser for the DSL (e.g., grid 6x4, row 1: E W S W E, etc.). Construction of an intermediate representation (IR) for the façade grid and rule set.
Middle-end: Static analyses for layout validation (e.g., “doors must be on ground floor”, “windows evenly spaced”). Transform passes, such as automatic symmetry expansion and run-length compaction of empty cells.
Back-end: SVG code generation to visualize the compiled façade. JSON code generation for machine-readable CV or generative design workflows.
I will incrementally develop and test each compiler stage, using small façade programs to verify correctness.

**How will you empirically measure success?**
Functional correctness: Run 5+ façade input programs that cover both valid and invalid cases. Verify that the compiler detects all rule violations and produces valid SVG/JSON outputs for correct inputs.
Rule enforcement accuracy: Check that each rule (e.g., window spacing ≥ threshold) is enforced precisely and reported with useful error messages.
Output quality and efficiency: Confirm that rendered SVGs visually match intended façade patterns.
Optionally, I can measure extensibility by adding one new rule (e.g., “no window above door”) to demonstrate modular compiler design.

**Team members:**
Me

### PR DESCRIPTION
**Pull request by itingtsai**
closes #618

### CODE REVIEW

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 1]
```diff
@@ -0,0 +1,409 @@
++++
+title = "Facade Layout Compiler (# 618)"
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: No need for the issue number here. Remember that this blog post is for public consumption; the public doesn't care about GitHub internals.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+[extra]
+bio = """
+  I-Ting Tsai is a CS PhD student at Cornell University.
+"""
+[[extra.authors]]
+name = "I-Ting Tsai"
++++
+
+# Experience Report: Facade Layout Compiler
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: You don't need a heading here; your blog post already has a title.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+## Alignment with Proposal
+
+My original proposal outlined three key deliverables:
+
+1. **Custom DSL for facade specification:** Implemented with support for row declarations, repeat expressions, compact syntax, comments, and rule declarations
+2. **Static rule enforcement:** Distinct semantic checks with error/warning classification
+3. **Empirical validation through testing:** pass a comprehensive test suite
+
+The scope evolved slightly during implementation. More details are described in the sections below.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md:31]: Again, remember that this blog post will be published publicly. To the extent that you actually want to talk about the original proposal, you'll want to link to it.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+---
+
+## What did you do?
+
+The complete implementation is available in [facade_compiler.py](https://github.com/itingtsai/Facade_Layout_Compiler/blob/main/facade_compiler.py)
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]:
```suggestion
The complete implementation is available in [facade_compiler.py](https://github.com/itingtsai/Facade_Layout_Compiler/blob/main/facade_compiler.py).
```

But I'm not sure you need both this and the other link to your repo at the top of the post.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+### DSL Design
+
+The facade DSL is designed as a minimal, domain-specific language for describing building facades as discrete 2D grids. Each facade is expressed as a set of numbered row declarations, where each row contains a sequence of symbolic elements:
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md:39]: Maybe this is a silly question, but why include the row numbers? Do they have any specific meaning? Would it work to just leave them off?

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+
+5. **Case insensitivity:** Both `E` and `e` are valid (normalized to uppercase internally).
+
+6. **Rule declarations:** Optional rule declarations (e.g., `rule min_window_spacing: 2`) parameterize semantic checks. Rules are stored in the IR and consulted during analysis.
+
+7. **Auto-fill behavior:** Rows may have different lengths; the compiler automatically pads shorter rows with empty cells to create a rectangular grid. Auto-filled cells are tracked in metadata so users can distinguish compiler-inferred content from explicit specification.
+
+### Design Rationale
+
+The DSL intentionally avoids geometric coordinates, numeric dimensions, or continuous values. By focusing on symbolic structure, the language:
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: What does "symbolic structure" mean?

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+
+- Simplifies parsing and semantic analysis
+- Enables clear error reporting at the language level
+- Allows the same IR to be reused for multiple back-end targets (SVG, JSON, or potentially image generation prompts)
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: I don't really understand how these advantages are connected to omitting those features listed above. These all seem like perfectly reasonable things to implement for any input language.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+
+The grammar is regular and unambiguous, making it amenable to simple hand-written parsing without requiring a parser generator.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: Overall, I don't think I'm clear on what the point of this section is—you have already discussed the design of the syntax, and this doesn't add much to the motivation you've already done.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+
+**Token types recognized:**
+
+- Keywords: `row`, `rule`
+- Symbols: `E`, `W`, `S`, `D`, `C`
+- Integers (for row numbers and repeat counts)
+- Punctuation: `:` (colon), `*` (multiply for repeats)
+- Identifiers (for rule names and symbol sequences)
+
+**Key design decisions:**
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: I don't think these are "design decisions." They are standard attributes of any lexer.

Sorry if this is too blunt, but were these "key design decisions" sections written by an LLM? If so, I don't think you should include them. One sentence of your own insights would be worth much more than many bullet points of LLM-generated description.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+
+- **Invalid character detection:** The lexer explicitly rejects characters not in the DSL's alphabet (punctuation, Unicode, emoji). This catches input errors immediately rather than propagating malformed tokens to the parser.
+
+- **Negative number handling:** Rather than tokenizing `-` as a separate operator, the lexer detects the pattern `-[digit]` and reports "Negative numbers not allowed" directly. This provides clearer error messages than a generic parse error.
+
+- **Symbol sequence recognition:** When the lexer encounters a multi-character word like `EWCWE`, it checks if all characters are valid symbols. If so, it emits a single `IDENTIFIER` token that the parser later expands. This enables compact syntax without complicating the grammar.
+
+- **Error handling:** Lexical errors are reported but do not prevent continued scanning. This allows the compiler to report multiple errors in a single pass, improving the user experience during iterative development.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: I'm a little uncomfortable with the "marketing" language used in some of these features, such as "improving the user experience during iterative development." I don't think this is a real advantage—the "user" in this case is just you! So it's a little weird to be highlighting this as a major contribution of the lexer—it distracts from the point, which is just that you have a working lexer that generates the tokens you need.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+
+### Parser Implementation
+
+The parser is a hand-written top-down predictive parser that consumes the token stream and constructs an AST consisting of:
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: What does "predictive" mean in this context?

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+
+- **Program:** The root node containing rules and rows
+- **RuleDecl:** A rule declaration with name and value(s)
+- **RowDecl:** A row with its number and list of cells
+- **Cell:** A single symbol at a source location
+- **RepeatExpr:** A symbol with a repeat count
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: The names here don't match the names in the EBNF grammar (which use `snake_case`).

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+rule_decl   ::= 'rule' IDENTIFIER ':' rule_value+
+row_decl    ::= 'row' INTEGER ':' cell+
+cell        ::= SYMBOL | repeat_expr | symbol_sequence
+repeat_expr ::= SYMBOL '*' INTEGER
+```
+
+**Note on Grammar Classification:** This grammar is regular, and the structure is purely sequential without nesting. The language could equivalently be expressed as a regular expression:
+
+```
+(rule IDENTIFIER : value+)* (row INTEGER : (SYMBOL | SYMBOL * INTEGER)+)+
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: Is it `value` or `rule_value`?

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+- **RuleDecl:** A rule declaration with name and value(s)
+- **RowDecl:** A row with its number and list of cells
+- **Cell:** A single symbol at a source location
+- **RepeatExpr:** A symbol with a repeat count
+
+**Grammar (EBNF):**
+
+```
+program     ::= rule_decl* row_decl+
+rule_decl   ::= 'rule' IDENTIFIER ':' rule_value+
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md:113]: `rule_value` seems to not be part of the grammar?

File: content/blog/2025-12-16-facade-layout-compiler.md (hunk 13 above)
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: Is the `*` in `SYMBOL * INTEGER` supposed to be a token? Then maybe you want to use `'*'`?

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+
+**Note on Grammar Classification:** This grammar is regular, and the structure is purely sequential without nesting. The language could equivalently be expressed as a regular expression:
+
+```
+(rule IDENTIFIER : value+)* (row INTEGER : (SYMBOL | SYMBOL * INTEGER)+)+
+```
+
+**Note on Parser Classification:** The parser follows a top-down, predictive parsing strategy with one function per major grammar production (`parse()`, `parse_rule()`, `parse_row()`, `parse_cells()`). It is not strictly a recursive-descent parser because no function calls itself—the grammar's repetition operators (`*` and `+`) are implemented with iterative `while` loops rather than recursive calls. This iterative approach is natural for a regular grammar where recursion is unnecessary. The parser uses single-token lookahead via `check()` to predict which production to apply, making it a predictive parser.
+
+**Key design decisions:**
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: Again, I don't think these are "key design decisions." They are just a few very detailed notes about low-level implementation details.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+
+**Step 2: Normalize grid dimensions**
+
+After construction, the grid may be jagged (rows of different lengths). The normalizer:
+
+- Computes the maximum row width
+- Pads shorter rows with `E` (empty) cells
+- Records which cells were auto-filled in metadata
+
+This guarantees a rectangular grid for all downstream processing.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: This seems to more or less repeat what you've said above about how padding works. Is it possible to say all this in 1 or 2 sentences?

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+
+### Architectural Rule Checks
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md:147]: I think this entire section would be a lot more readable if you made it shorter. Just a simple bulleted list of all the rules you implement would suffice! The longer sections here don't add much beyond a simple, straightforward description of the rule.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+
+The analyzer compares each row with its horizontal mirror. Asymmetric rows produce warnings to inform the designer of potential stylistic inconsistency. This is informational rather than prescriptive.
+
+### Error Classification
+
+I deliberately distinguished between errors (fatal violations) and warnings (stylistic concerns):
+
+- **Error:** Structurally invalid facade. Compilation halts; no output generated.
+- **Warning:** Stylistic issue or optional constraint violation. Compilation continues; warning recorded in output.
+- **Info:** Informational message (e.g., auto-fill notification). No effect on compilation.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: These are the standard definitions of these terms in any logging setup. I don't think you need to define them.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+
+### SVG Output
+
+The SVG generator produces a visual representation where each cell is rendered as a colored rectangle with a label. Cells are grouped by type in the SVG DOM, making it easy to style or manipulate specific element types.
+
+### PNG Rasterization
+
+For environments that don't support SVG, the compiler can produce PNG output via CairoSVG.
+
+---
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: Can you please include at least one example of the visual output? Your compiler produces images that are hopefully interesting to look at—it would be very sad if this post didn't include any!

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+- Is `@` an invalid character (lexical) or an invalid symbol (semantic)?
+- Is `row 1: X E E` a syntax error or a semantic error?
+
+I resolved this by establishing boundaries:
+
+- **Lexical errors:** Characters not in the DSL's alphabet (punctuation, Unicode, emoji)
+- **Syntax errors:** Malformed constructs (missing colons, invalid repeat expressions, missing row numbers)
+- **Semantic errors:** Valid syntax with invalid meaning (floating doors, disconnected chimneys, invalid symbol letters like `X`)
+
+The key insight was that `X` is a valid identifier syntactically but not a valid symbol semantically. The parser accepts it and emits a semantic error during cell processing.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: This seems like a long way to say: "errors can be emitted at different parts of the compiler." I'm not sure what the insight is here. It's also not clear *why* you decided that `X` should be "a valid identifier syntactically but not a valid symbol semantically"; other choices also seem like they could work.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+Initially, I tried a simple positional check (e.g., chimney cells must be on the top row). This failed for common patterns like diagonal chimneys:
+
+```
+row 1: C E E
+row 2: C C E
+row 3: E C C
+row 4: E E C
+```
+
+This chimney is valid. It connects to the roof through a diagonal path. The flood-fill approach correctly handles this by treating chimneys as connected components rather than isolated cells.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: This example doesn't seem to have a roof in it… maybe that's intentional (it's just a chimney in isolation?), but it is a little confusing that this says it "connects to the roof."

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 22]
```diff
@@ -0,0 +1,409 @@
...
+
+### 3. Zero-Count Repeat Expressions
+
+The expression `E*0` presented an interesting design decision. Options:
+
+1. Treat as syntax error (reject)
+2. Treat as single cell (emit one `E`)
+3. Treat as zero cells (emit nothing)
+
+I chose option 3 because it's the most mathematically consistent interpretation and enables useful patterns like conditional exclusion. The implementation simply doesn't add any cell to the AST when count is 0.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: I think you've already covered this thoroughly above when you first introduced the feature.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+### 5. Error Recovery and Multiple Error Reporting
+
+Fixing one error often revealed another, requiring multiple compile-fix cycles.
+
+I implemented error recovery in both the lexer and parser:
+
+- **Lexer:** Reports invalid characters but continues scanning
+- **Parser:** Uses `sync_to_row()` to skip to the next row declaration after errors
+
+This required careful state management to ensure the compiler doesn't report spurious follow-on errors while still catching genuine independent errors.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: What does that "careful state management" consist of? What kind of state are you referring to?

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 24]
```diff
@@ -0,0 +1,409 @@
...
+
+---
+
+## Were you successful?
+
+### Quantitative Evaluation
+
+The testing results printout is available in [test.txt](https://github.com/itingtsai/Facade_Layout_Compiler/blob/main/test/test.txt)
+
+Comprehensive testing is hard. There is no existing benchmark. After implementing and validating each new rule, I use an LLM for test ideation (e.g., "Give me edge cases to test user inputs containing characters outside the defined DSL elements.") and then manually determine and verify the expected behavior. This approach helped broaden test coverage to corner cases that I wouldn't have considered. The categorized test suite made it easy to verify that each feature worked in isolation and in combination.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: You have already said, just above, that you "use an LLM for test ideation."

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 25]
```diff
@@ -0,0 +1,409 @@
...
+| Stress | 5 | Stress tests (large grids, many rows) |
+| Recovery | 3 | Error recovery |
+| Combo | 4 | Feature combinations |
+| ZeroRepeat | 9 | Zero-count repeat expressions |
+
+All 146 tests pass, demonstrating:
+
+- **Functional correctness:** The compiler accepts valid programs and rejects invalid ones
+- **Rule enforcement precision:** Architectural constraints are enforced exactly as specified
+- **Robustness:** The compiler handles syntactic variations, malformed input, and edge cases gracefully
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: Please delete this list (for the same "marketing" reasons as above). You can just say that the test pass.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 26]
```diff
@@ -0,0 +1,409 @@
...
+
+### Qualitative Validation
+
+I output 5 facade examples (json, svg, png) together with image gen visualizations which are available in the [example folder](https://github.com/itingtsai/Facade_Layout_Compiler/tree/main/example).
+
+Beyond automated testing, I validated output quality through visual inspection:
+
+- **SVG correctness:** Generated SVGs accurately represent the input grid with correct colors and positioning
+- **Error message quality:** Error messages include source locations and actionable hints
+- **Warning appropriateness:** Symmetry and spacing warnings provide useful feedback without blocking compilation
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: There is again some "marketing" language here that makes me uncomfortable. For instance, "actionable" and "useful" are a hard-to-defend claim. Also, this paragraph suggests a kind of distance that you don't really have in your evaluation: it seems to suggest that you were judging the compiler's output from first principles and thinking carefully about how useful it was. But of course, you wrote those error messages! It seems impossible to judge the helpfulness of messages you wrote for yourself—or rather, you would never write a message that you yourself found unhelpful, so this isn't really falsifiable.

Long story short: I don't think this can be considered part of a qualitative evaluation, because nothing was systematically assessed.

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 27]
```diff
@@ -0,0 +1,409 @@
...
+
+I conducted an experiment using an image generation model (Nano Banana) to produce photorealistic facade images from compiled layouts. The workflow:
+
+1. Compile facade DSL → JSON + SVG (+ PNG)
+2. Construct prompt from grid structure and style description
+3. Generate image using multimodal model
+
+**Text prompt format:** *Generate a [-style] facade of a [low-rise/mid-rise/high-rise] building based on this image, where W = window, D = door, S = spacing, E = empty, and C = chimney.*
+
+The results were correct most of the time. The model understood the general layout but sometimes struggled with precise spatial positioning. This suggests the compiled output could serve as input for generative design, though current models may need additional guidance.
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md:353]: What was this experiment attempting to assess?

Also, if you generated an image with this approach, it would again be sad not to include it here in this blog post!

**Review by sampsyo — CHANGES_REQUESTED** [review]
Very glad that this project panned out pretty much exactly as planned!

I am concerned about a few places in the writeup that seem to be longer and more "marketing-y" than is really necessary—in a few places, a sentence or two would suffice where there is currently a section of somewhat "fluffy" extrapolation. As I said in one inline comment, I apologize if this is overstepping, but these segments feel like there were LLM-written? I don't object to your using an LLM during some step in the writing process, but they notoriously produce very long, content-free text where a few short sentences would be better.

When you have a chance to address the inline writing suggestions, can you please use GitHub's "request review" button to let me know when you're ready to publish?

**Comment by itingtsai** [comment]
Hi Adrian, thank you very much for the detailed feedback. Sorry for using an LLM to help polish the writing flow, which ended up making some sections overly verbose and repetitive. I’ve revised the content based on your feedback and responded directly to your questions inline. Please let me know if anything still feels off. Thanks again for your guidance and patience!

File: content/blog/2025-12-16-facade-layout-compiler.md [hunk 28]
```diff
@@ -0,0 +1,397 @@
...
+row 1: E E E E E E E E C E E 
+row 2: S S S S S S S S C S S 
+row 3: S S W W S S S W W S S 
+row 4: S S W W S S S W W S S 
+row 5: S S S S S S S S S S S 
+row 6: E 
+row 6: S S D D S S W W W W S
+row 7: S S D D S S W W W W S 
+```
+![Facade 1](https://github.com/itingtsai/Facade_Layout_Compiler/blob/main/example/facade_1/facade_1.svg)
```
Comment (sampsyo) [inline content/blog/2025-12-16-facade-layout-compiler.md]: I unfortunately don't think this image embedding works, because this is a link to a GitHub page and not the actual image data itself. I can think of two possible ways to resolve this:

* Use a URL to the raw data, like <https://raw.githubusercontent.com/itingtsai/Facade_Layout_Compiler/refs/heads/main/example/facade_1/facade_1.svg>.
* You can include the SVG files directly with your blog post, in this repository, by making your post a directory here. See my README <https://github.com/sampsyo/cs6120>. Namely, the paragraph beginning "To include images or other resources in your post…"

**Review by sampsyo — CHANGES_REQUESTED** [review]
Looking pretty good overall! Unfortunately the current image links are broken—please see my inline suggestion for how to fix this.

File: content/blog/2025-12-16-facade-layout-compiler.md (hunk 28 above)
Comment (itingtsai) [reply content/blog/2025-12-16-facade-layout-compiler.md]: Hi Adrian, thanks for catching this. I’ve updated them to use raw URLs. Please let me know if there are still issues. Sorry for the back and forth. Thank you again for your time!

Comment (sampsyo) [reply content/blog/2025-12-16-facade-layout-compiler.md]: Looks pretty good from here! Let's publish and see if it works.
